src/libstd/io/buffered.rs

```diff
@@ -0,0 +1,637 @@
+    }
+}
+
+#[stable(feature = "rust1", since = "1.0.0")]
+impl<R> fmt::Debug for BufReader<R> where R: fmt::Debug {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "BufReader {{ reader: {:?}, buffer: {}/{} }}",
+               self.inner, self.buf.position(), self.buf.get_ref().len())
+    }
+}
+
+/// Wraps a Writer and buffers output to it
+///
+/// It can be excessively inefficient to work directly with a `Write`. For
+/// example, every call to `write` on `TcpStream` results in a system call. A
+/// `BufWriter` keeps an in memory buffer of data and writes it to the
+/// underlying `Write` in large, infrequent batches.
+///
+/// This writer will be flushed when it is dropped.
+pub struct BufWriter<W> {
+    inner: Option<W>,
+    buf: Vec<u8>,
+}
+
+impl<W: Write> BufWriter<W> {
+    /// Creates a new `BufWriter` with a default buffer capacity
+    pub fn new(inner: W) -> BufWriter<W> {
+        BufWriter::with_capacity(DEFAULT_BUF_SIZE, inner)
+    }
+
+    /// Creates a new `BufWriter` with the specified buffer capacity
+    pub fn with_capacity(cap: usize, inner: W) -> BufWriter<W> {
+        BufWriter {
+            inner: Some(inner),
+            buf: Vec::with_capacity(cap),
+        }
+    }
+
+    fn flush_buf(&mut self) -> io::Result<usize> {
+        let len = self.buf.len();
+        let n = try!(self.inner.as_mut().unwrap().write(&self.buf[]));
+        // FIXME would be better expressed as .remove(0..n) if it existed
+        unsafe {
+            ptr::copy_memory(self.buf.as_mut_ptr(),
+                             self.buf.as_ptr().offset(n as isize),
+                             len - n);
+        }
+        self.buf.truncate(len - n);
+        Ok(n)
+    }
+
+    /// Gets a reference to the underlying writer.
+    pub fn get_ref(&self) -> &W { self.inner.as_ref().unwrap() }
+
+    /// Gets a mutable reference to the underlying write.
+    ///
+    /// # Warning
+    ///
+    /// It is inadvisable to directly read from the underlying writer.
+    pub fn get_mut(&mut self) -> &mut W { self.inner.as_mut().unwrap() }
+
+    /// Unwraps this `BufWriter`, returning the underlying writer.
+    ///
+    /// The buffer is flushed before returning the writer.
+    pub fn into_inner(mut self) -> W {
+        // FIXME(#12628): is panicking the right thing to do if flushing panicks?
+        self.flush_buf().unwrap();
+        self.inner.take().unwrap()
+    }
+}
+
+impl<W: Write> Write for BufWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.buf.len() + buf.len() > self.buf.capacity() {
+            try!(self.flush_buf());
+        }
+        if self.buf.len() == 0 && buf.len() >= self.buf.capacity() {
+            self.inner.as_mut().unwrap().write(buf)
+        } else {
+            let amt = cmp::min(buf.len(), self.buf.capacity());
+            Write::write(&mut self.buf, &buf[..amt])
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
```

Comment:
>/// Flush this output stream, ensuring that all intermediately buffered contents reach their destination.

This implementation is wrong because flush_buf does not ensure this.